saml/lib/Saml.php

```diff
@@ -135,6 +135,59 @@ function processLoginResponse($request)
             );
         }
 
+        if ($this->config->isUserGroupsSettingEnabled()) {
+            $samlUserGroupsSetting = $this->config->getUserGroupsSetting();
+            if (isset($samlUserGroupsSetting['attribute']) && $samlUserGroupsSetting['attribute'] != ''
+                && isset($samlUserGroupsSetting['separator']) && $samlUserGroupsSetting['separator'] != ''
+                && isset($attributes[$samlUserGroupsSetting['attribute']])) {
+
+                $allGroups = array();
+                foreach(jAcl2DbUserGroup::getGroupList() as $group) {
+                    if ($group->grouptype == 2) {
+                        // private group, let's ignore
+                        continue;
+                    }
+
+                    $allGroups[] = $group->id_aclgrp;
+                }
+
+                $groupsOfUser = array_map('trim', explode($samlUserGroupsSetting['separator'], $attributes[$samlUserGroupsSetting['attribute']]));
+                if (isset($samlUserGroupsSetting['prefix']) && $samlUserGroupsSetting['prefix'] != '') {
+                    $prefix = $samlUserGroupsSetting['prefix'];
+                    $groupsOfUser = array_map(
+                        function($g) use ($prefix) {
+                            return substr($g, strlen($prefix));
+                        },
+                        array_filter($groupsOfUser, function($g) use ($prefix) {
+                            return substr($g, 0, strlen($prefix)) == $prefix;
+                        })
+                    );
+                }
+
+                $groupsToRemove = array_filter($allGroups, function($g) use ($groupsOfUser) {
+                    return !array_search($g, $groupsOfUser);
+                });
+                $groupsToAdd = array_filter($groupsOfUser,function($g) use ($allGroups) {
+                    return !array_search($g, $allGroups);
+                });
+
```

Comment:
The algorithm could be improved. Some lizmap instances may have dozen groups, but users are often only into one or two groups. And then, here, it will try to remove the user from dozen groups although the user is not in these groups. It will make dozen useless SQL queries.

The lists of groups to add or remove, should be constructed from the current list of groups of the user, not from the list of all groups (but it also should check if groups to add, are existing).

Reply:
I have not well understood `jAcl2DbUserGroup::getGroupList()`. I have improved the algorithm, with login groups.